pkg/routeagent_driver/handlers/ovn/non_gateway_route_controller.go

```diff
@@ -0,0 +1,147 @@
+	connectionHandler      *ConnectionHandler
+	remoteSubnets          sets.Set[string]
+	stopCh                 chan struct{}
+	transitSwitchIP        string
+}
+
+//nolint:gocritic // Ignore hugeParam
+func NewNonGatewayRouteController(config watcher.Config, connectionHandler *ConnectionHandler,
+	k8sClientSet clientset.Interface, namespace string,
+) (*NonGatewayRouteController, error) {
+	// We'll panic if config is nil, this is intentional
+	var err error
+
+	controller := &NonGatewayRouteController{
+		connectionHandler: connectionHandler,
+		remoteSubnets:     sets.New[string](),
+	}
+
+	config.ResourceConfigs = []watcher.ResourceConfig{
+		{
+			Name:         "NonGatewayRoute watcher",
+			ResourceType: &submarinerv1.NonGatewayRoute{},
+			Handler: watcher.EventHandlerFuncs{
+				OnCreateFunc: controller.nonGatewayRouteCreatedOrUpdated,
+				OnUpdateFunc: controller.nonGatewayRouteCreatedOrUpdated,
+				OnDeleteFunc: controller.nonGatewayRouteDeleted,
+			},
+			SourceNamespace: namespace,
+		},
+	}
+
+	node, err := nodeutil.GetLocalNode(k8sClientSet)
+	if err != nil {
+		return nil, errors.Wrap(err, "error getting the node")
+	}
+
+	annotations := node.GetAnnotations()
+
+	transitSwitchIP := annotations["k8s.ovn.org/node-transit-switch-port-ifaddr"]
+	if transitSwitchIP == "" {
+		logger.Infof("No transit switch IP configured on node %q", node.Name)
+		return controller, nil
+	}
+
+	controller.transitSwitchIP, err = jsonToIP(transitSwitchIP)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error parsing transit switch IP")
+	}
+
+	controller.nonGatewayRouteWatcher, err = watcher.New(&config)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating resource watcher")
+	}
+
+	err = controller.nonGatewayRouteWatcher.Start(controller.stopCh)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error starting non gateway route controller")
+	}
+
+	logger.Infof("Started NonGatewayRouteController")
+
+	return controller, nil
+}
+
+func (g *NonGatewayRouteController) nonGatewayRouteCreatedOrUpdated(obj runtime.Object, _ int) bool {
+	submNonGWRoute := obj.(*submarinerv1.NonGatewayRoute)
+
+	err := g.reconcileRemoteSubnets(submNonGWRoute, true)
+	if err != nil {
+		logger.Errorf(err, "Error creating or updating router policies for remote subnet %q", g.remoteSubnets)
+		return true
+	}
+
+	return false
+}
+
+func (g *NonGatewayRouteController) nonGatewayRouteDeleted(obj runtime.Object, _ int) bool {
+	submNonGWRoute := obj.(*submarinerv1.NonGatewayRoute)
+
+	err := g.reconcileRemoteSubnets(submNonGWRoute, false)
+	if err != nil {
+		logger.Errorf(err, "Error deleting policies for remote subnet %q", g.remoteSubnets)
+		return true
+	}
+
+	return false
+}
+
+func (g *NonGatewayRouteController) reconcileRemoteSubnets(submNonGWRoute *submarinerv1.NonGatewayRoute, addSubnet bool) error {
+	if len(submNonGWRoute.RoutePolicySpec.NextHops) == 0 {
+		logger.Warningf("The NonGatewayRoute does not have next hop %v", submNonGWRoute)
```

Comment:
Please add a comment when we would hit this case.

Reply:
Done.